Cannot invoke "org.eclipse.jetty.io.RetainableByteBuffer.getByteBuffer()" because "buffer" is null

**Jetty version(s)**

12.0.16

**Jetty Environment**

ee8

**Java version/vendor** `(use: java -version)`

openjdk version "22.0.2" 2024-07-16
OpenJDK Runtime Environment Temurin-22.0.2+9 (build 22.0.2+9)
OpenJDK 64-Bit Server VM Temurin-22.0.2+9 (build 22.0.2+9, mixed mode, sharing)

**OS type/version**

Linux 6.1.0-1015-oem #15-Ubuntu SMP PREEMPT_DYNAMIC Fri Jun 16 09:51:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**Description**

We register a consistent rate of NPEs on multiple servers running jetty since at least Dec 2024. They surface when serializing JSON responses with Jackson. The response size is < 3kB.

```
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.io.RetainableByteBuffer.getByteBuffer()" because "buffer" is null
        at org.eclipse.jetty.io.ArrayByteBufferPool.release(ArrayByteBufferPool.java:281)
        at org.eclipse.jetty.io.ArrayByteBufferPool$Buffer.release(ArrayByteBufferPool.java:653)
        at org.eclipse.jetty.io.ArrayByteBufferPool.removeAndRelease(ArrayByteBufferPool.java:244)
        at org.eclipse.jetty.ee8.nested.HttpOutput.lockedReleaseBuffer(HttpOutput.java:599)
        at org.eclipse.jetty.ee8.nested.HttpOutput.onWriteComplete(HttpOutput.java:310)
        at org.eclipse.jetty.ee8.nested.HttpOutput.close(HttpOutput.java:573)
        at org.eclipse.jetty.io.WriteThroughWriter.close(WriteThroughWriter.java:103)
        at org.eclipse.jetty.ee8.nested.ResponseWriter.close(ResponseWriter.java:136)
        ...

Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.io.RetainableByteBuffer.getByteBuffer()" because "buffer" is null
        at org.eclipse.jetty.io.ArrayByteBufferPool.release(ArrayByteBufferPool.java:281)
        at org.eclipse.jetty.io.ArrayByteBufferPool$Buffer.release(ArrayByteBufferPool.java:653)
        at org.eclipse.jetty.io.ArrayByteBufferPool.removeAndRelease(ArrayByteBufferPool.java:244)
        at org.eclipse.jetty.ee8.nested.HttpOutput.lockedReleaseBuffer(HttpOutput.java:599)
        at org.eclipse.jetty.ee8.nested.HttpOutput.onWriteComplete(HttpOutput.java:310)
        at org.eclipse.jetty.ee8.nested.HttpOutput.write(HttpOutput.java:781)
        at java.base/java.io.ByteArrayOutputStream.writeTo(ByteArrayOutputStream.java:170)
        at org.eclipse.jetty.io.WriteThroughWriter$Utf8Writer.append(WriteThroughWriter.java:335)
        at org.eclipse.jetty.io.WriteThroughWriter.write(WriteThroughWriter.java:133)
        at org.eclipse.jetty.ee8.nested.ResponseWriter.write(ResponseWriter.java:174)
        ...
```

**How to reproduce?**

Not sure yet. Flagging the issue in case somebody with context about Jetty can help pin down the problem.

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
@@ -308,13 +308,10 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
     {
         bucket.recordRelease();
 
-        RetainableByteBuffer buffer = entry.getPooled();
-        if (buffer == null)
-        {
-            // Removed before being released, record removal.
-            bucket.recordRemove();
+        if (entry.isTerminated())
             return;
-        }
+
+        RetainableByteBuffer buffer = entry.getPooled();
         BufferUtil.reset(buffer.getByteBuffer());
 
         // Release the buffer and check the memory 1% of the times.

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
@@ -34,7 +34,6 @@ import java.util.concurrent.atomic.LongAdder;
 import java.util.function.IntUnaryOperator;
 import java.util.function.ToLongFunction;
 import java.util.stream.Collectors;
-
 import org.eclipse.jetty.io.internal.CompoundPool;
 import org.eclipse.jetty.io.internal.QueuedPool;
 import org.eclipse.jetty.util.BufferUtil;
@@ -308,7 +307,11 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
 
         RetainableByteBuffer buffer = entry.getPooled();
         if (buffer == null)
+        {
+            // Removed before being released, record removal.
+            bucket.recordRemove();
             return;
+        }
         BufferUtil.reset(buffer.getByteBuffer());
 
         // Release the buffer and check the memory 1% of the times.

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
@@ -34,6 +34,7 @@ import java.util.concurrent.atomic.LongAdder;
 import java.util.function.IntUnaryOperator;
 import java.util.function.ToLongFunction;
 import java.util.stream.Collectors;
+
 import org.eclipse.jetty.io.internal.CompoundPool;
 import org.eclipse.jetty.io.internal.QueuedPool;
 import org.eclipse.jetty.util.BufferUtil;
@@ -256,7 +257,9 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
     {
         RetainableByteBuffer actual = buffer;
         while (actual instanceof RetainableByteBuffer.Wrapper wrapper)
+        {
             actual = wrapper.getWrapped();
+        }
 
         if (actual instanceof ReservedBuffer reservedBuffer)
         {
@@ -319,7 +322,7 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
         if (entry.release())
         {
             if (used % 100 == 0)
-               checkMaxMemory(bucket, buffer.isDirect());
+                checkMaxMemory(bucket, buffer.isDirect());
             return;
         }
 
@@ -448,7 +451,9 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
     {
         long size = 0;
         for (RetainedBucket bucket : direct ? _direct : _indirect)
+        {
             size += count.applyAsLong(bucket) * bucket.getCapacity();
+        }
         return size;
     }
 

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java
@@ -307,6 +307,8 @@ public class ArrayByteBufferPool implements ByteBufferPool, Dumpable
         bucket.recordRelease();
 
         RetainableByteBuffer buffer = entry.getPooled();
+        if (buffer == null)
+            return;
         BufferUtil.reset(buffer.getByteBuffer());
 
         // Release the buffer and check the memory 1% of the times.

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
@@ -203,6 +203,7 @@ public class QueuedPool<P> implements Pool<P>
         // val/false -> idle
         // val/true  -> in use
         private final AtomicMarkableReference<P> pooled = new AtomicMarkableReference<>(null, false);
+        private P removedReference; // Only written before setting 'pooled', only read after getting 'pooled'.
 
         private QueuedEntry(QueuedPool<P> pool)
         {
@@ -243,7 +244,8 @@ public class QueuedPool<P> implements Pool<P>
         @Override
         public P getPooled()
         {
-            return pooled.getReference();
+            P reference = pooled.getReference();
+            return reference != null ? reference : removedReference;
         }
 
         void acquire()
@@ -273,6 +275,7 @@ public class QueuedPool<P> implements Pool<P>
             P p = pooled.get(inUse);
             if (p == null && inUse[0])
                 return false;
+            removedReference = p;
             pooled.set(null, true);
             return true;
         }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
@@ -203,7 +203,6 @@ public class QueuedPool<P> implements Pool<P>
         // val/false -> idle
         // val/true  -> in use
         private final AtomicMarkableReference<P> pooled = new AtomicMarkableReference<>(null, false);
-        private P removedReference; // Only written before setting 'pooled', only read after getting 'pooled'.
 
         private QueuedEntry(QueuedPool<P> pool)
         {
@@ -214,96 +213,139 @@ public class QueuedPool<P> implements Pool<P>
         public boolean enable(P pooled, boolean acquire)
         {
             Objects.requireNonNull(pooled);
-            boolean[] inUse = new boolean[1];
-            P p = this.pooled.get(inUse);
-            if (p != null)
-            {
-                if (pool.isTerminated())
-                    return false;
-                throw new IllegalStateException("Entry already enabled " + this + " for " + pool);
-            }
-            if (inUse[0])
-                return false; // terminated
 
-            this.pooled.set(pooled, acquire);
-            if (acquire)
+            while (true)
             {
-                if (pool.isTerminated())
+                boolean[] state = new boolean[1];
+                P p = this.pooled.get(state);
+                if (p != null)
                 {
-                    this.pooled.set(null, false);
-                    return false;
+                    if (pool.isTerminated())
+                        return false;
+                    throw new IllegalStateException("Entry already enabled " + this + " for " + pool);
                 }
-                return true;
-            }
-            else
-            {
-                return pool.requeue(this);
+                if (state[0])
+                    return false; // terminated
+
+                if (!this.pooled.compareAndSet(null, pooled, false, acquire))
+                    continue;
+
+                if (acquire)
+                {
+                    if (pool.isTerminated())
+                    {
+                        this.pooled.set(null, true);
+                        return false;
+                    }
+                    return true;
+                }
+                else
+                {
+                    return pool.requeue(this);
+                }
             }
         }
 
         @Override
         public P getPooled()
         {
-            P reference = pooled.getReference();
-            return reference != null ? reference : removedReference;
+            return  pooled.getReference();
         }
 
         void acquire()
         {
-            boolean[] inUse = new boolean[1];
-            P p = pooled.get(inUse);
-            if (p == null || inUse[0])
-                return; // terminated
-            pooled.set(p, true);
+            while (true)
+            {
+                boolean[] state = new boolean[1];
+                P p = pooled.get(state);
+                boolean idle = isIdle(p, state[0]);
+                if (!idle)
+                    return;
+                if (pooled.compareAndSet(p, p, false, true))
+                    return;
+            }
         }
 
         @Override
         public boolean release()
         {
-            boolean[] inUse = new boolean[1];
-            P p = pooled.get(inUse);
-            if (p == null || !inUse[0])
-                return false;
-            pooled.set(p, false);
-            return pool.requeue(this);
+            while (true)
+            {
+                boolean[] state = new boolean[1];
+                P p = pooled.get(state);
+                boolean inUse = isInUse(p, state[0]);
+                if (!inUse)
+                    return false;
+                if (pooled.compareAndSet(p, p, true, false))
+                    return pool.requeue(this);
+            }
         }
 
         @Override
         public boolean remove()
         {
-            boolean[] inUse = new boolean[1];
-            P p = pooled.get(inUse);
-            if (p == null && inUse[0])
-                return false;
-            removedReference = p;
-            pooled.set(null, true);
-            return true;
+            while (true)
+            {
+                boolean[] state = new boolean[1];
+                P p = pooled.get(state);
+                boolean terminated = isTerminated(p, state[0]);
+                if (terminated)
+                    return false;
+                if (pooled.compareAndSet(p, null, state[0], true))
+                    return true;
+            }
         }
 
         @Override
         public boolean isReserved()
         {
-            return pooled.getReference() == null;
+            boolean[] state = new boolean[1];
+            P p = pooled.get(state);
+            return isReserved(p, state[0]);
+        }
+
+        private static boolean isReserved(Object item, boolean state)
+        {
+            return item == null && !state;
         }
 
         @Override
         public boolean isIdle()
         {
-            return !pooled.isMarked();
+            boolean[] state = new boolean[1];
+            P p = pooled.get(state);
+            return isIdle(p, state[0]);
+        }
+
+        private static boolean isIdle(Object item, boolean state)
+        {
+            return item != null && !state;
         }
 
         @Override
         public boolean isInUse()
         {
-            return pooled.isMarked();
+            boolean[] state = new boolean[1];
+            P p = pooled.get(state);
+            return isInUse(p, state[0]);
+        }
+
+        private static boolean isInUse(Object item, boolean state)
+        {
+            return item != null && state;
         }
 
         @Override
         public boolean isTerminated()
         {
-            boolean[] inUse = new boolean[1];
-            P p = pooled.get(inUse);
-            return p == null && inUse[0];
+            boolean[] state = new boolean[1];
+            P p = pooled.get(state);
+            return isTerminated(p, state[0]);
+        }
+
+        private static boolean isTerminated(Object item, boolean state)
+        {
+            return item == null && state;
         }
     }
 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
@@ -249,7 +249,7 @@ public class QueuedPool<P> implements Pool<P>
         @Override
         public P getPooled()
         {
-            return  pooled.getReference();
+            return pooled.getReference();
         }
 
         void acquire()

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/QueuedPool.java
@@ -214,9 +214,9 @@ public class QueuedPool<P> implements Pool<P>
         {
             Objects.requireNonNull(pooled);
 
+            boolean[] state = new boolean[1];
             while (true)
             {
-                boolean[] state = new boolean[1];
                 P p = this.pooled.get(state);
                 if (p != null)
                 {
@@ -254,9 +254,9 @@ public class QueuedPool<P> implements Pool<P>
 
         void acquire()
         {
+            boolean[] state = new boolean[1];
             while (true)
             {
-                boolean[] state = new boolean[1];
                 P p = pooled.get(state);
                 boolean idle = isIdle(p, state[0]);
                 if (!idle)
@@ -269,9 +269,9 @@ public class QueuedPool<P> implements Pool<P>
         @Override
         public boolean release()
         {
+            boolean[] state = new boolean[1];
             while (true)
             {
-                boolean[] state = new boolean[1];
                 P p = pooled.get(state);
                 boolean inUse = isInUse(p, state[0]);
                 if (!inUse)
@@ -284,9 +284,9 @@ public class QueuedPool<P> implements Pool<P>
         @Override
         public boolean remove()
         {
+            boolean[] state = new boolean[1];
             while (true)
             {
-                boolean[] state = new boolean[1];
                 P p = pooled.get(state);
                 boolean terminated = isTerminated(p, state[0]);
                 if (terminated)

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
@@ -13,6 +13,7 @@
 
 package org.eclipse.jetty.io;
 
+import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -20,6 +21,7 @@ import java.util.List;
 import java.util.Map;
 
 import org.eclipse.jetty.io.internal.CompoundPool;
+import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.util.ConcurrentPool;
 import org.eclipse.jetty.util.Pool;
 import org.hamcrest.Matchers;
@@ -60,6 +62,9 @@ public class ArrayByteBufferPoolTest
         for (int i = 0; i < bufferCount; i++)
         {
             RetainableByteBuffer rbb = pool.acquire(1, false);
+            ByteBuffer byteBuffer = rbb.getByteBuffer();
+            BufferUtil.flipToFill(byteBuffer);
+            byteBuffer.position(1);
             rbbs.add(rbb);
         }
         rbbs.forEach(pool::removeAndRelease);
@@ -67,9 +72,13 @@ public class ArrayByteBufferPoolTest
         for (RetainableByteBuffer rbb : rbbs)
         {
             assertThat(rbb.getByteBuffer(), notNullValue());
+            assertEquals(1, rbb.getByteBuffer().position());
         }
+
+        assertEquals(0L, pool.getAvailableHeapByteBufferCount());
+
         String dump = pool.dump();
-        assertThat(dump, containsString("[capacity=4096,in-use=0/0,pooled/acquires/releases=512/1024/1024(50.000%),non-pooled/evicts/removes=0/0/1024]"));
+        assertThat(dump, containsString("[capacity=4096,in-use=0/0,pooled/acquires/releases=512/1024/1024(50.000%),non-pooled/evicts/removes=0/0/512]"));
     }
 
     @Test

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
@@ -42,7 +42,7 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 public class ArrayByteBufferPoolTest
 {
     @Test
-    public void testRemoveAndReleaseCompoundPool()
+    public void testRemoveAndReleaseFromCompoundPool()
     {
         int bufferCount = ConcurrentPool.OPTIMAL_MAX_SIZE * 2;
         List<RetainableByteBuffer> rbbs = new ArrayList<>();
@@ -61,8 +61,12 @@ public class ArrayByteBufferPoolTest
             RetainableByteBuffer rbb = pool.acquire(1, false);
             rbbs.add(rbb);
         }
-
         rbbs.forEach(pool::removeAndRelease);
+
+        for (RetainableByteBuffer rbb : rbbs)
+        {
+            assertThat(rbb.getByteBuffer(), notNullValue());
+        }
     }
 
     @Test

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
@@ -18,6 +18,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+
 import org.eclipse.jetty.io.internal.CompoundPool;
 import org.eclipse.jetty.util.ConcurrentPool;
 import org.eclipse.jetty.util.Pool;

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
@@ -18,7 +18,6 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
-
 import org.eclipse.jetty.io.internal.CompoundPool;
 import org.eclipse.jetty.util.ConcurrentPool;
 import org.eclipse.jetty.util.Pool;
@@ -48,6 +47,7 @@ public class ArrayByteBufferPoolTest
         List<RetainableByteBuffer> rbbs = new ArrayList<>();
 
         ArrayByteBufferPool pool = new ArrayByteBufferPool();
+        pool.setStatisticsEnabled(true);
         for (int i = 0; i < bufferCount; i++)
         {
             RetainableByteBuffer rbb = pool.acquire(1, false);
@@ -67,6 +67,8 @@ public class ArrayByteBufferPoolTest
         {
             assertThat(rbb.getByteBuffer(), notNullValue());
         }
+        String dump = pool.dump();
+        assertThat(dump, containsString("[capacity=4096,in-use=0/0,pooled/acquires/releases=512/1024/1024(50.000%),non-pooled/evicts/removes=0/0/1024]"));
     }
 
     @Test

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/ArrayByteBufferPoolTest.java
@@ -42,6 +42,30 @@ import static org.junit.jupiter.api.Assertions.assertTrue;
 public class ArrayByteBufferPoolTest
 {
     @Test
+    public void testRemoveAndReleaseCompoundPool()
+    {
+        int bufferCount = ConcurrentPool.OPTIMAL_MAX_SIZE * 2;
+        List<RetainableByteBuffer> rbbs = new ArrayList<>();
+
+        ArrayByteBufferPool pool = new ArrayByteBufferPool();
+        for (int i = 0; i < bufferCount; i++)
+        {
+            RetainableByteBuffer rbb = pool.acquire(1, false);
+            rbbs.add(rbb);
+        }
+        rbbs.forEach(Retainable::release);
+        rbbs.clear();
+
+        for (int i = 0; i < bufferCount; i++)
+        {
+            RetainableByteBuffer rbb = pool.acquire(1, false);
+            rbbs.add(rbb);
+        }
+
+        rbbs.forEach(pool::removeAndRelease);
+    }
+
+    @Test
     public void testDump()
     {
         ArrayByteBufferPool pool = new ArrayByteBufferPool(0, 10, 100, Integer.MAX_VALUE, 200, 200);

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/internal/QueuedPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/internal/QueuedPoolTest.java
@@ -83,10 +83,10 @@ public class QueuedPoolTest
 
         Pool.Entry<String> e1 = pool.acquire();
 
+        assertThat(e1.release(), is(true));
         assertThat(e1.remove(), is(true));
         assertThat(e1.remove(), is(false));
-        assertThat(e1.release(), is(false));
-        assertThat(e1.getPooled(), notNullValue());
+        assertThat(e1.getPooled(), nullValue());
     }
 
     @Test
@@ -99,7 +99,7 @@ public class QueuedPoolTest
         assertThat(e1.remove(), is(true));
         assertThat(e1.remove(), is(false));
         assertThat(e1.release(), is(false));
-        assertThat(e1.getPooled(), notNullValue());
+        assertThat(e1.getPooled(), nullValue());
     }
 
     @Test

--- a/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/internal/QueuedPoolTest.java
+++ b/jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/internal/QueuedPoolTest.java
@@ -76,6 +76,20 @@ public class QueuedPoolTest
     }
 
     @Test
+    public void testReleaseBeforeRemove()
+    {
+        Pool<String> pool = new QueuedPool<>(1);
+        pool.reserve().enable("aaa", false);
+
+        Pool.Entry<String> e1 = pool.acquire();
+
+        assertThat(e1.remove(), is(true));
+        assertThat(e1.remove(), is(false));
+        assertThat(e1.release(), is(false));
+        assertThat(e1.getPooled(), notNullValue());
+    }
+
+    @Test
     public void testRemoveBeforeRelease()
     {
         Pool<String> pool = new QueuedPool<>(1);
@@ -85,6 +99,7 @@ public class QueuedPoolTest
         assertThat(e1.remove(), is(true));
         assertThat(e1.remove(), is(false));
         assertThat(e1.release(), is(false));
+        assertThat(e1.getPooled(), notNullValue());
     }
 
     @Test

